tests/foreman/api/test_discoveredhost.py

```diff
@@ -67,561 +77,484 @@ def _create_discovered_host(name=None, ipaddress=None, macaddress=None):
-    """Implements tests for foreman discovery feature"""
 
-    def _assertdiscoveredhost(self, hostname, user_config=None):
-        """Check if host is discovered and information about it can be
-        retrieved back
+def _assert_discovered_host(host, channel=None, user_config=None):
+    """Check if host is discovered and information about it can be
+    retrieved back
 
-        Introduced a delay of 300secs by polling every 10 secs to get expected
-        host
-        """
-        default_config = entity_mixins.DEFAULT_SERVER_CONFIG
-        for _ in range(30):
-            discovered_host = entities.DiscoveredHost(user_config or default_config).search(
-                query={'search': f'name={hostname}'}
-            )
-            if discovered_host:
-                break
-            time.sleep(10)
-        else:
-            raise HostNotDiscoveredException(
-                f"The host {hostname} is not discovered within 300 seconds"
-            )
-        return discovered_host[0]
-
-    @classmethod
-    @skip_if_not_set('vlan_networking')
-    def setUpClass(cls):
-        """Steps to Configure foreman discovery
-
-        1. Build PXE default template
-        2. Create Organization/Location
-        3. Update Global parameters to set default org and location for
-           discovered hosts.
-        4. Enable auto_provision flag to perform discovery via discovery
-           rules.
-        """
-        super().setUpClass()
-
-        # Build PXE default template to get default PXE file
-        entities.ProvisioningTemplate().build_pxe_default()
-        # let's just modify the timeouts to speed things up
-        ssh.command(
-            "sed -ie 's/TIMEOUT [[:digit:]]\\+/TIMEOUT 1/g' "
-            "/var/lib/tftpboot/pxelinux.cfg/default"
+    Introduced a delay of 300secs by polling every 10 secs to get expected
+    host
+    """
+    # assert that server receives DHCP discover from hosts PXELinux
+    for pattern in [
+        (
+            f"DHCPDISCOVER from {host.mac}",
+            "DHCPDISCOVER",
+        ),
+        (f"DHCPACK on [0-9.]+ to {host.mac}", "DHCPACK"),
+    ]:
+        try:
+            dhcp_pxe = _wait_for_log(channel, pattern[0], timeout=10)
+        except TimedOutError:
+            # raise assertion error
+            raise AssertionError(f'Timed out waiting for {pattern[1]} from VM')
+
+    groups = re.search('DHCPACK on (\\d.+) to', dhcp_pxe.out)
+    assert len(groups.groups()) == 1, 'Unable to parse bootloader ip address'
+    pxe_ip = groups.groups()[0]
+
+    # assert that server retrieves pxelinux config over TFTP
+    for pattern in [
+        (f'Client {pxe_ip} finished pxelinux.0', 'pxelinux.0'),
+        (f'Client {pxe_ip} finished pxelinux.cfg/default', 'pxelinux.cfg/default'),
+        (f'Client {pxe_ip} finished boot/fdi-image/vmlinuz0', 'fdi-image/vmlinuz0'),
+        (f'Client {pxe_ip} finished boot/fdi-image/initrd0.img', 'fdi-image/initrd0.img'),
```

Comment:
Hm, could you get away with only iterating over file names, and constructing `pattern[0]` inside loop? Looks like they all follow same pattern, you only need to pass `pxe_ip` and file name itself.

Reply:
I certainly could, but I didn't want to sacrifice readability. I think this way anybody could easily see what we're up for.
What do you think?